packages/jest-runtime/src/index.ts

```diff
@@ -737,12 +740,108 @@ export default class Runtime {
     ) {
       return this.loadEsmModule(resolved, query);
     }
-
+    if (skipCjsModules) {
+      return null;
+    }
     return this.loadCjsAsEsm(referencingIdentifier, resolved, context);
   }
 
+  private async checkForCyclicDependencies(
+    module: VMModule,
+    parentChain: Set<VMModule>,
+  ): Promise<boolean> {
+    if (
+      module.status !== 'unlinked' ||
+      this._hasCyclicDependencyMap.has(module)
+    ) {
+      return this._hasCyclicDependencyMap.get(module) as boolean;
+    }
+
+    // if the dependency resolves to a CJS module, it will be undefined
+    const dependencies: Array<VMModule | undefined> = await Promise.all(
+      (module.dependencySpecifiers || []).map(async (modSpec: VMModule) => {
+        const dependency = await this.resolveModule(
+          modSpec,
+          module.identifier,
+          module.context,
+          true,
+        );
+        return dependency;
+      }),
+    );
+
+    let hasCyclicDependency = false;
+    await Promise.all(
+      dependencies.filter(Boolean).map(async (possiblyCyclicMod: VMModule) => {
+        const isCyclic = parentChain.has(possiblyCyclicMod);
+        if (isCyclic) {
+          // if the found dependency is already in the parentChain
+          // this is a cyclic dependency. As such, everything between where
+          // that module is in the parentChain and the end of the chain
+          // (e.g., the current module) is also cyclic.
+          const chainInOrder = [...parentChain];
+          const indexOf = chainInOrder.indexOf(possiblyCyclicMod);
+          for (const cyclicDep of chainInOrder.slice(indexOf)) {
+            this._hasCyclicDependencyMap.set(cyclicDep, true);
+          }
+          hasCyclicDependency = true;
+        } else {
+          this.checkForCyclicDependencies(
+            possiblyCyclicMod,
+            new Set([...parentChain, possiblyCyclicMod]),
+          );
+        }
+      }),
+    );
+    // if somehow the map got set to false for this module, but we now know it's true
+    // set it to true
+    if (!this._hasCyclicDependencyMap.has(module) || hasCyclicDependency) {
+      this._hasCyclicDependencyMap.set(module, hasCyclicDependency);
+    }
+    return hasCyclicDependency;
+  }
+
+  private linkModule(
+    specifier: string,
+    referencingModule: VMModule,
+    parentChain: Set<VMModule>,
+  ): VMModule {
+    const preserveLoadOrder =
+      this._config.preserveLoadOrder || this._globalConfig?.preserveLoadOrder;
+
+    // ensure that every import fully evaluates before any siblings are allowed to import.
+    let linkPromiseChain = (Promise<VMModule | void>).resolve();
+    if (preserveLoadOrder) {
```

Comment:
this seems a bit more advanced than it needs to. Is there any reason not to unwrap this with something like

```diff
diff --git i/packages/jest-runtime/src/index.ts w/packages/jest-runtime/src/index.ts
index f4ff5238a7..154cd877bb 100644
--- i/packages/jest-runtime/src/index.ts
+++ w/packages/jest-runtime/src/index.ts
@@ -801,42 +801,32 @@ export default class Runtime {
     return hasCyclicDependency;
   }
 
-  private linkModule(
+  private async linkModule(
     specifier: string,
     referencingModule: VMModule,
     parentChain: Set<VMModule>,
-  ): VMModule {
+  ): Promise<VMModule> {
+    const mod = await this.resolveModule<VMModule>(
+      specifier,
+      referencingModule.identifier,
+      referencingModule.context,
+    );
+
     const preserveLoadOrder =
       this._config.preserveLoadOrder || this._globalConfig?.preserveLoadOrder;
 
-    // ensure that every import fully evaluates before any siblings are allowed to import.
-    let linkPromiseChain = (Promise<VMModule | void>).resolve();
     if (preserveLoadOrder) {
-      linkPromiseChain = linkPromiseChain.then(async () => {
-        const mod = await this.resolveModule<VMModule>(
-          specifier,
-          referencingModule.identifier,
-          referencingModule.context,
-        );
-        const hasCyclicDependency = await this.checkForCyclicDependencies(
-          mod,
-          parentChain,
-        );
-        if (mod.status === 'unlinked' && !hasCyclicDependency) {
-          await this.linkAndEvaluateModule(
-            mod,
-            new Set<VMModule>([...parentChain, mod]),
-          );
-        }
-        return mod;
-      });
-      return linkPromiseChain;
+      // ensure that every import fully evaluates before any siblings are allowed to import.
+      const hasCyclicDependency = await this.checkForCyclicDependencies(
+        mod,
+        parentChain,
+      );
+      if (mod.status === 'unlinked' && !hasCyclicDependency) {
+        await this.linkAndEvaluateModule(mod, new Set([...parentChain, mod]));
+      }
     }
-    return this.resolveModule<VMModule>(
-      specifier,
-      referencingModule.identifier,
-      referencingModule.context,
-    );
+
+    return mod;
   }
 
   private async linkAndEvaluateModule(
```
?

Reply:
I don't think it can be made simpler - the `linkPromiseChain` is necessary:

```
import './file1.js';
import './file2.js';
```

without it, these two will call `linkAndEvaluateModule` in parallel. If `file2.js` depends on something was setup by `file1.js`, it won't work. Similarly, `resolveModule` must be called within the chain because the issue is with cjs vs esm modules - so something setup by esm and depended on by cjs must be available at the time `resolveModule` is called - since, in the case of cjs, `resolveModule` also evaluates.

Reply:
the first check in `checkForCyclicDependencies` is for `mod.status !== 'unlinked'` - so all we'd save is the cost of the call itself